Fixed loading/embedding fonts and custom save function

elizabethmcd opened the pull request:
This PR reworks the logic of loading our custom fonts. The `load_arcadia_fonts()` function now uses the `extrafont` package instead of `showtext` package in order to embed the fonts in exported plots. This fixes the large problem in #27 where fonts loaded with `showtext` were exported as paths/shapes and not as text or vectors, and therefore not editable in Illustrator. This is now fixed. The font loading function has also been changed to execute and check/load the fonts when the package is loaded instead of being called every time the `theme_arcadia()` function is applied to a plot. This PR also adds a custom save function which addresses #5 so users don't have to remember things like DPI and sizing (which I'm not sure I have correctly here but it is easily adjustable). 

This should be the last PR before I cut the first release. There are still some small issues and enhancement requests but what was fixed here were previously the largest barriers to cutting a release. 

@taylorreiter I kindly request that in addition to reviewing this PR, that you install this branch of the package and test that this actually works on your system, and that you can export plots with editable text in Illustrator. This would give me increased confidence that things are working properly and ready for a first release of the package.

taylorreiter commented on R/onAttach.R:
Oh this is an interesting approach. Did you find it documented/detailed anywhere? I'd love to read more about it!

elizabethmcd replied:
I've seen it a couple times in scripts called `aaaa.R` or `zzz.R`, and I found that to be a little confusing in case other want to develop new things for the package later so I called the script just what it does for now. I mainly saw it here for inspiration: https://github.com/wch/extrafont/blob/master/R/aaa.r. I originally tried `.onLoad` but that had a lot of problems and .`onAttach` seems to work fine.

taylorreiter reviewed (commented):
One question inline, but when I tried to install and run this I get:
```
library(arcadiathemeR)
Loading Suisse fonts...
Custom fonts 'Suisse Int'l Semi Bold, Suisse Int'l Medium, Suisse Int'l Mono' not found. Using fallback font 'sans'.
```

I scrolled through the readme and didn't see any instructions for where my font files should live for `arcadiathemeR` to find them.

elizabethmcd commented:
> One question inline, but when I tried to install and run this I get:
> 
> ```
> library(arcadiathemeR)
> Loading Suisse fonts...
> Custom fonts 'Suisse Int'l Semi Bold, Suisse Int'l Medium, Suisse Int'l Mono' not found. Using fallback font 'sans'.
> ```
> 
> I scrolled through the readme and didn't see any instructions for where my font files should live for `arcadiathemeR` to find them.

I just added instructions for downloading the fonts (specifically for the R package to work you must have the `.ttf` formatted files) and where to put them. I also added the note that this is only confirmed to work on Mac OS.

mezarque commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,44 @@
+  # Ensure the orientation argument is valid
+  orientation <- match.arg(orientation)
+
+  # Set size based on orientation
```
In the future, it would be worthwhile to have these match the standard panel sizes in the style guide.

elizabethmcd replied:
Should I name them `full_wide`, `float_wide` etc?

elizabethmcd replied:
I went ahead and changed this to `panel_size` and the options/sizes match those in the style guide in pixels. I was uncertain about the height measurements though because the style guide only seems to give widths

mezarque replied:
Yeah, I think we should keep things consistent. We use those terms in the `arcadia-pycolor` implementation, e.g. [here](https://github.com/Arcadia-Science/arcadia-pycolor/blob/6f940d815d6195972bb4bdbe2544c3894098b128/arcadia_pycolor/style_defaults.py#L15C1-L30C1). 

```python
# Common figure sizes for Arcadia Creative Cloud library templates.
FULL_WIDE = (13.33333, 5.27777)
FULL_SQUARE = (6.52777, 6.52777)
FLOAT_WIDE = (9.16667, 5.27777)
FLOAT_SQUARE = (4.44444, 4.44444)
HALF_SQUARE = (6.38888, 6.38888)

# Dictionary of common figure sizes.
FIGURE_SIZES = {
    "full_wide": FULL_WIDE,
    "full_square": FULL_SQUARE,
    "float_wide": FLOAT_WIDE,
    "float_square": FLOAT_SQUARE,
    "half_square": HALF_SQUARE,
}
```

mezarque replied:
Note that the numbers for sizes above are at 72dpi:

13.33333 * 72 ~= 960px, +2 * 20px padding around plots = 1000px wide for `FULL_WIDE`

mezarque commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,44 @@
+    filename = filepath,
+    plot = plot,
+    device = "pdf",
+    dpi = 300,
```
How does this dpi affect the relative sizing of fonts? When opening the PDFs in Illustrator, do you see the right font sizes still? I had to save at 72 dpi in order to get things to look correct in `matplotlib`.

elizabethmcd replied:
I think the font sizes look ok? I can change to 72 DPI if that's what you have.

elizabethmcd replied:
I changed to 72 DPI, it doesn't look different to me I think?

mezarque replied:
I think the real test would be to try to drag and drop the plots into the current Illustrator templates and see whether they fit. Not critical to get this done by Tuesday though.

mezarque reviewed (commented):
Left a couple of questions!

elizabethmcd commented:
@mezarque @taylorreiter Wondering if I can get one of you to take one last look and raise any last issues or give approval so I can cut a release today before teaching the AUG session.

mezarque commented on R/arcadia.R:
```diff
@@ -23,9 +23,7 @@
-theme_arcadia <- function(x_axis_type = "numerical", y_axis_type = "numerical", background = TRUE) {
-  # load all Suisse fonts and store
-  loaded_suisse_fonts <- load_arcadia_fonts()
+theme_arcadia <- function(x_axis_type = "numerical", y_axis_type = "numerical", background = TRUE, padding = c(.25, .25, .25, .25)) {
```
Maybe extract the padding as a constant?

mezarque commented on R/load_fonts.R:
```diff
@@ -8,23 +8,24 @@
 load_arcadia_fonts <- function(custom_font = "Suisse", fallback_font = "sans") {
-  showtext::showtext_auto(enable = TRUE)
+  # define font names to check for
+  font_names <- c("Suisse Int'l", "Suisse Int'l Semi Bold", "Suisse Int'l Medium", "Suisse Int'l Mono")
```
Should this use / import the constants from `R/arcadia.R`?

elizabethmcd replied:
I'm not sure how to export constants across scripts, so for now I'll leave it like this and make an issue. Unless I have to make a specific function like `define_constants()` somewhere and export it so that it's available to all other scripts.

mezarque commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,62 @@
+#' p <- ggplot(mtcars, aes(x = hp, y = mpg)) + geom_point()
+#' save_arcadia_plot(p, "plot.pdf", "full_wide")
+#' }
+save_arcadia_plot <- function(filepath, plot, panel_size = c("full_wide", "float_wide", "half_square", "full_square", "float_square")) {
```
To avoid duplication, you might pull out the list of panel size names into a constant, and unpack the list as needed later (e.g. when erroring on invalid panel size).

mezarque commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,62 @@
+  panel_size <- tryCatch(match.arg(panel_size), error = function(e) NA)
+
+  if (is.na(panel_size)) {
+    stop("Invalid panel_size. Please choose from: 'full_wide', 'float_wide', 'half_square', 'full_square', 'float_square'.")
```
I think choosing a panel size should be optional. Not all data will fit neatly into the pre-determined panel sizes. In `arcadia_pycolor`, we separate the panel size designation from the save function; users can use the panel sizes if they want to, but are free to choose other sizes if they want. The panel sizes are just to make things simpler for those who want to use the Creative Cloud templates.

mezarque replied:
Don't have to resolve this by tomorrow though; feel free to make an issue.

elizabethmcd replied:
Previously I had `landscape` and `portrait` as options, so I could just add those back in where those were just the default 8x11 or 11x8

mezarque replied:
That seems like a reasonable solution for now!

mezarque commented on README.Rmd:
````diff
@@ -21,6 +21,8 @@ You can install arcadiathemeR from Github using the `remotes` package. If you do
 remotes::install_github("Arcadia-Science/arcadiathemeR")
 ```
 
+To use the custom fonts you need to download the `TTF` formatted font files and place in the `Users/YOURUSERNAME/Library/Fonts/` directory. Check out the Arcadia Science Brand Assets page in Notion to find these. This should only need to be performed once even if the package is updated over time. These steps and functionality have only been confirmed to work on Mac OS.
````
Might want to suggest that users download and double-click on the fonts in order to let FontBook handle the installation on macOS. I think that would be easier for most users to understand.

taylorreiter replied:
I think it might also be nice to call out what font will be fallen back to (so users know they can still use the package if they don't have the fonts)

elizabethmcd replied:
addressed

mezarque reviewed (commented):
Recommended some new constants and added a couple of thoughts.

taylorreiter commented on DESCRIPTION:
```diff
@@ -11,12 +11,12 @@ Encoding: UTF-8
 Roxygen: list(markdown=TRUE)
 RoxygenNote: 7.3.1
+Depends:
+    R (>= 4.0)
```
just curious, where does this version dep come from?

elizabethmcd replied:
I added it since it seems to be good practice to have a minimum R version to depend on. I thought at minimum 4.0 would be good since R is at 4.4 and RStudio requires at least 3.6.

taylorreiter commented on R/arcadia.R:
```diff
@@ -23,9 +23,7 @@
-theme_arcadia <- function(x_axis_type = "numerical", y_axis_type = "numerical", background = TRUE) {
-  # load all Suisse fonts and store
-  loaded_suisse_fonts <- load_arcadia_fonts()
+theme_arcadia <- function(x_axis_type = "numerical", y_axis_type = "numerical", background = TRUE, padding = c(.25, .25, .25, .25)) {
```
It appears as though padding isnt' documented as a param, but I may be missing something.

elizabethmcd replied:
addressed

taylorreiter commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,62 @@
+    filename = filepath,
+    plot = plot,
+    device = "pdf",
+    dpi = 72,
```
is 72 dpi the right dpi? that feels very low, but maybe it doesn't matter bc pdfs are vectorized?

elizabethmcd replied:
I followed from what Dennis has in arcadia-pycolor, but I think he had to use 72 DPI for some of the sizings to work out. I did some tests with 300 DPI as well and don't think I could see a difference with the PDFs.

mezarque replied:
72 dpi is the correct resolution for web content; we can explore in the future what settings make the most sense for print / projected media!

taylorreiter commented on R/save_arcadia_plot.R:
```diff
@@ -0,0 +1,62 @@
+  )
+}
+
+# Example usage
+# library(ggplot2)
+# p <- ggplot(mtcars, aes(x = hp, y = mpg)) + geom_point()
+# save_arcadia_plot("plot_full_wide.pdf", p, "full_wide")
+# save_arcadia_plot("plot_float_wide.pdf", p, "float_wide")
+# save_arcadia_plot("plot_half_square.pdf", p, "half_square"
+# save_arcadia_plot("plot_full_square.pdf", p, "full_square")
+# save_arcadia_plot("plot_float_square.pdf", p, "float_square")
```
is this mean to be commented out instead of provided as function documentation?

elizabethmcd replied:
fixed

taylorreiter commented on README.Rmd:
```diff
@@ -53,6 +55,19 @@ ggplot(data=diamonds, aes(x=cut, fill=cut)) +
+  theme(legend.position = "bottom")
+
+save_arcadia_plot("man/figures/arcadia-plot.pdf", plot, panel_size = "full_square")
+
```
extra space

elizabethmcd replied:
addressed

taylorreiter reviewed (approved):
Approving for the sake of deadling, but if Mert has time to take a look that might be nice. Please also address or file issues for Dennis's comments before merging (which it looks like you were already doing)
